### ISSUE TITLE
Which training loss to use?

### ISSUE THREAD

**Issue by braceal** [ramanathanlab/molecular_dynamics_neural_operator#2]
Hi @zongyi-li, I see there are a few training losses calculated here. Which one should be used for backpropagation? Thanks!

https://github.com/ramanathanlab/molecular_dynamics_neural_operator/blob/main/graph_kernel.py#L247

**Comment by zongyi-li** [comment]
We usually use the relative l2 loss. It has some regularization effect and usually does better than mse.

**Comment by braceal** [comment]
Thanks for getting back to us on this! We have the network running on our data, however, we need help understanding what the output of the model represents.  In the original script you had, it looks like you were loading a solution matrix from disk, but we were not sure what that matrix represents or what it would be for our use case. In the current version we have, the network is outputting a (28, 1) dimensional tensor (batch size omitted here). This makes sense due to the fact that this system has 28 CA atoms in each graph. We set up the dataloader to batch graphs which contain the following object for each training example:
```
data = PairData(
    x=node_features, # A [28, 20] matrix of one hot encodings representing the type of amino acid
    edge_attr=edge_attr, # A (num_edges, 6) array containing positions [Ax, Ay, Az, Bx, By, Bz] for connected nodes A, B 
    edge_index_s=edge_index_s, # The adjacency matrix of the model input
    edge_index_t=edge_index_t, # A future target adjacency matrix that we would like to predict using the model
)
```

From this data, we can correctly pass in x, edge_attr, and edge_index_s into the model. However, we added the target adjacency matrix, edge_index_t, as a best guess for what the y solution vector might be. To compute the loss, we are using an inner product decoder taken from pytorch_geometric to map the (28,1) model output to a matrix which can be compared with edge_index_t. Unfortunately, this does not produce a reasonable loss curve. Can you help us to better understand what the model output represents and what the target y tensor should be in this use case?

Here is a link to the point where we call the loss function, the loss function is implemented in the same file:
https://github.com/ramanathanlab/molecular_dynamics_neural_operator/blob/main/graph_kernel.py#L326

Thanks!

**Comment by zongyi-li** [comment]
Thanks for running the model! If I understand correctly, we want to model the evolution of the position of the 28 nodes. In this case, the input tensor can also include the position, so x has the shape [28, 20+3]. 

And the output y will be the location at the next time step [28, 3].  And we can make the output of the model [28, 3] (line 266, make the output dim 3 instead of 1), and compute the loss between y and the output. Given the location y [28, 3], I think it's not hard to compute the [28,28] contact map, but we probably do not need to use the contact map in training. (later we can do the map from the previous 10 time steps to the next time steps).

if we have 100000 time steps in total, we can create 100000 such graphs and do batch training.

**Comment by braceal** [comment]
Ah, thank you so much! This makes perfect sense. I will try to implement it tomorrow and hopefully generate some results. 

Out of curiosity, do you think concatenating the onehot amino acid vector in R^20 with a R^3 dense fp32 position vector will work well? Or should we send the onehot representation through an nn.Embedding layer before concatenating with the R^3 vector?

And for the post-training model validation, is this the correct way to proceed: 
On held out data, use the model to generate the next step prediction y[t+1], and then recurrently feed the prediction back into the model to generate y[t+2], and so on. 

This way we could generate a sequence of predictions and then check it against the held out values. Is there a better way to do it?

**Comment by zongyi-li** [comment]
Yes, it is a wonderful idea to have an embedding layer for the amino vector. Feel free to modify the network architecture.

Regarding the time sequence, yes, for the current formulation we can do this. As the further steps, we may input longer time steps or use the RNN structure to propagate in time.

**Comment by braceal** [comment]
Hi @zongyi-li, I've applied all changes we discussed (embedding layer, predicting the future positions, and reverting back to the LpLoss). It seems to be working much better now, although there may be some overfitting. Can I ask how you normally split the dataset for training? I'm simply taking the first 80K points in the trajectory as the training set and the last 20K as validation and then shuffling the training set. Perhaps the validation set is too far off distribution from the training set, or perhaps there is standard overfitting. Also do you know which hyperparameters are most important for tuning? I can setup a hyperparameter optimization run to search the space of the more important ones. Here are the current loss values:

```
Epoch: 0	Time: 280.46989192813635	train_loss: inf	valid_loss: 19519186.648824643
Epoch: 1	Time: 279.99646987672895	train_loss: 37993252.73047379	valid_loss: 194774.91710609043
Epoch: 2	Time: 280.2342201722786	train_loss: 1771614.3854684217	valid_loss: 158743.9725995186
Epoch: 3	Time: 281.06537519115955	train_loss: 238890.94583907494	valid_loss: 354895.3505578652
Epoch: 4	Time: 279.3349408050999	train_loss: 114876.06826492456	valid_loss: 158163.64719723433
Epoch: 5	Time: 279.09397740475833	train_loss: 30723.698311720138	valid_loss: 28337.599349144177
Epoch: 6	Time: 278.9476942215115	train_loss: 1550.9020732977451	valid_loss: 28602.108530582525
Epoch: 7	Time: 278.98410066682845	train_loss: 968.2141553805425	valid_loss: 31604.36552404746
Epoch: 8	Time: 279.11566322576255	train_loss: 153.79098599996323	valid_loss: 31643.805104622475
Epoch: 9	Time: 282.95680757798254	train_loss: 149.7726386265877	valid_loss: 31698.29372024536
Epoch: 10 Time: 278.42053725011647	train_loss: 144.25925045747024	valid_loss: 31763.32479066115
Epoch: 11 Time: 277.65680337976664	train_loss: 137.6228583653768	valid_loss: 31859.34394498972
Epoch: 12 Time: 278.03175760246813	train_loss: 130.1678433418274	valid_loss: 31859.580012297018
```

**Comment by braceal** [comment]
Just another update, it seems like the loss got very large but is decreasing again:
```
Epoch: 12	Time: 278.03175760246813	train_loss: 130.1678433418274	valid_loss: 31859.580012297018
Epoch: 13	Time: 277.98594562523067	train_loss: 129.20975808608227	valid_loss: 31859.980809627436
Epoch: 14	Time: 277.7079010633752	train_loss: 129.17784467110266	valid_loss: 31860.293252944946
Epoch: 15	Time: 279.1066285762936	train_loss: 278462123994668.4	valid_loss: 24205248676.102566
Epoch: 16	Time: 279.47745225951076	train_loss: 17662586552.615383	valid_loss: 11601962194.051283
Epoch: 17	Time: 279.272941400297	train_loss: 9172323000.205128	valid_loss: 6931343415.794871
Epoch: 18	Time: 279.603521768935	train_loss: 5903563029.948718	valid_loss: 4975044747.48718
Epoch: 19	Time: 279.27379039675	train_loss: 3893380819.282051	valid_loss: 4185405824.820513
Epoch: 20	Time: 279.2390311518684	train_loss: 3776754200.5128207	valid_loss: 2869571697.230769
Epoch: 21	Time: 278.6931283036247	train_loss: 2071235879.0769231	valid_loss: 2010211874.871795
Epoch: 22	Time: 278.69325015693903	train_loss: 1685390845.5384614	valid_loss: 2295208454.5641026
Epoch: 23	Time: 278.92515126708895	train_loss: 2513444540.3333335	valid_loss: 1417781729.8461537
Epoch: 24	Time: 278.8408908843994	train_loss: 1301714606.3333333	valid_loss: 796167246.1538461
Epoch: 25	Time: 278.7688510948792	train_loss: 639667098.8205128	valid_loss: 570305585.6410257
Epoch: 26	Time: 279.4855691026896	train_loss: 438481732.974359	valid_loss: 417599877.025641
```

Any ideas what may cause this behavior? Thanks!

**Comment by zongyi-li** [comment]
Thanks for getting this running. 

First, just to make sure the metric is right. When using (line 404) ` loss_fn = LpLoss(size_average=False)`, we didn't average among the batch. So, to get the real loss, we should be divided by the total number of training instances and testing instances (at line 334 and 349). The relative l2 loss should be somewhere between 0 and 1. (if we output a 0 vector, the loss should be 1).

It seems the training loss is jumping around, maybe we should use a smaller learning rate.

If the overfitting is large, maybe we should use a smaller mode, by decreasing the width of the kernel network.

**Comment by braceal** [comment]
Add another nn_conv layer and decrease the depth, also decrease the kernel_width to 128.

**Comment by braceal** [comment]
We can also try adding more timesteps to the input so the model can try to understand the velocities given the position time series.
